kore/src/Kore/Unification/UnifierImpl.hs

```diff
@@ -90,103 +90,55 @@ simplifyAnds (NonEmpty.sort -> patterns) = do
-    return (substitution' <> simplified)
-
-{- | Simplify a substitution so that each variable occurs only once.
-
-Simplify a conjunction of substitutions,
-
-@
-x₁ = t₁ ∧ ... ∧ xₙ = tₙ
-@
-
-where some of the @xᵢ@ may be the same so that each variable occurs on the
-left-hand side of only one substitution clause. New substitutions may be
-produced during simplification; @deduplicateSubstitution@ recurses until the
-solution stabilizes.
-
-See also: 'deduplicateOnce'
-
- -}
 deduplicateSubstitution
     :: forall variable unifier
     .   ( SimplifierVariable variable
         , MonadUnify unifier
         , WithLog LogMessage unifier
         )
-    => Substitution variable
-    -> unifier (Predicate variable)
+    =>  Substitution variable
+    ->  unifier
+            ( Syntax.Predicate variable
+            , Map (UnifiedVariable variable) (TermLike variable)
+            )
 deduplicateSubstitution =
-    worker . Predicate.fromSubstitution
+    worker Syntax.Predicate.makeTruePredicate . Substitution.toMultiMap
   where
-    isSingleton (_, _ :| duplicates) = null duplicates
-
-    insertSubstitution
-        :: forall variable1 term
-        .  Ord variable1
-        => Map variable1 (NonEmpty term)
-        -> (variable1, term)
-        -> Map variable1 (NonEmpty term)
-    insertSubstitution multiMap (variable, termLike) =
-        let push = (termLike :|) . maybe [] Foldable.toList
-        in Map.alter (Just . push) variable multiMap
-
-    worker conditional@Conditional { predicate, substitution }
-      | Substitution.isNormalized substitution || all isSingleton substitutions
-      = return conditional
+    worker
+        ::  Syntax.Predicate variable
+        ->  Map (UnifiedVariable variable) (NonEmpty (TermLike variable))
+        ->  unifier
+                ( Syntax.Predicate variable
+                , Map (UnifiedVariable variable) (TermLike variable)
+                )
+    worker predicate substitutions
+      | Just deduplicated <- traverse getSingleton substitutions
+      = return (predicate, deduplicated)
 
       | otherwise = do
-        deduplicated <- Foldable.fold <$> mapM deduplicateOnce substitutions
-        worker (Predicate.fromPredicate predicate <> deduplicated)
-
-      where
-        substitutions
-            :: [(UnifiedVariable variable, NonEmpty (TermLike variable))]
-        substitutions =
-            Map.toAscList
-            . Foldable.foldl' insertSubstitution Map.empty
-            . map sortRenamedVariable
-            $ Substitution.unwrap substitution
+        simplified <- collectConditions <$> traverse simplifyAnds substitutions
+        let -- Substitutions de-duplicated by simplification.
+            substitutions' = toMultiMap $ Conditional.term simplified
+            -- New conditions produced by simplification.
+            Conditional { predicate = predicate' } = simplified
+            predicate'' = Syntax.Predicate.makeAndPredicate predicate predicate'
```

Comment:
This line is kind of long (80 characters). I think it can be easily split up. The same for the line at 119.